app/src/dex/router/fill_route2.rs

```diff
@@ -0,0 +1,370 @@
+                        .phi
+                        .fill(current_value, &position.reserves)
+                        .expect("asset ids should match");
+
+                    assert_eq!(
+                        unfilled.amount,
+                        Amount::zero(),
+                        "unfilled amount for unconstrained frontier should be zero"
+                    );
+
+                    position.reserves = new_reserves;
+                    current_value = output;
+                    trace[i + 1].amount += output.amount;
+                }
+
+                tracing::debug!("filled input amount completely, breaking loop");
+                break 'filling;
+            };
+
+            // If there was a constraining position along the path, we want to
+            // completely consume its reserves, then work "outwards" along the
+            // path, propagating rounding errors forwards to the end of the path
+            // and backwards to the input.
+
+            // Example:
+            // 0     1     2     3      4         [trace index]
+            // UM => GM => GN => USD => ETH       [asset id]
+            //     0     1     2      3           [pair index]
+            //
+            // Suppose that pair 2 is the constraining pair, with 0.1 USD
+            // reserves.  To completely consume the 0.1 USD reserves, we need
+            // work backwards along the path to compute a sequence of input
+            // amounts that are valid trades to get to 0.1 USD output at pair 2,
+            // and work forwards to compute the corresponding output amounts at
+            // the end of the path.
+
+            // We want to completely consume the 0.1 USD reserves, then work backwards
+            // along the path to compute input amounts that are valid trades to get
+
+            let exactly_consumed_reserves = Value {
+                amount: frontier[constraining_index]
+                    .reserves_for(pairs[constraining_index].end)
+                    .expect("asset ids should match"),
+                asset_id: pairs[constraining_index].end,
+            };
+
+            tracing::debug!(
+                constraining_index,
+                exactly_consumed_reserves = ?exactly_consumed_reserves.amount,
+                "found constraining index, attempting to completely consume reserves"
+            );
+
+            // Work backwards along the path from the constraining position.
+            let mut current_value = exactly_consumed_reserves;
+            for i in (0..=constraining_index).rev() {
+                trace[i + 1].amount += current_value.amount;
+
+                let (new_reserves, prev_input) = frontier[i]
+                    .phi
+                    .fill_output(&frontier[i].reserves, current_value)
+                    .expect("asset ids should match")
+                    .expect("working backwards from most-constraining position should not exceed reserves");
```

Comment:
But the last constraint _is_ the most constraining one, because we know all the positions are distinct, and we apply each position's constraints, if any, as we pass through it.

If there were an earlier constraining position, we would have reduced the amount of output we got for the next hop, so the fact that we see a later position as constraining _that_ output means that the later position is more constraining than the earlier one.